<div class="text-cell-section-header">
<h1 style="text-align: center;"><strong>PROGRAMACI&Oacute;N PARA EL AN&Aacute;LISIS DE DATOS</strong></h1>
</div>
<h1 style="text-align: center;"><strong>Tarea 1 - Principios b&aacute;sicos de programaci&oacute;n cient&iacute;fica.</strong></h1>
<h1 style="text-align: center;"><strong>PRESENTADO POR: DARWIN MERCADO</strong></h1>
<h1 style="text-align: center;"><strong>GRUPO 203008069_3</strong></h1>
<h1 style="text-align: center;"><strong>PRESENTADO A FELIPE ALEXANDER PIPICANO GUZMAN</strong></h1>
<h1 style="text-align: center;"><strong>Universidad Nacional Abierta y a Distancia &ndash; UNAD</strong></h1>
<h1 style="text-align: center;"><strong>18 de septiembre del 2023</strong></h1>

# Presentación del ejercicio

# Ejercicio 1

Desarrolle códigos para ejecutar las siguientes tareas:

• Cargue en la variable df2 el conjunto de datos que se presentan
en el archivo CSV, Anexo 2 - Base de Datos CSV.csv, que se
encuentra en la carpeta de la Unidad 1 - Tarea 1 - Fundamentos
de programación científica con Python. Tenga en cuenta el tipo
de separador y los encabezados de las columnas.

• Cargue los datos Boston Housing Dataset que se encuentran en
la variable load_boston del paquete sklearn, en un dataframe de
pandas asignado a la variable df3.

• Construya un Panel con los tres (3) dataframes cargados
anteriormente: df1, df2 y df3.

• Programe una función que extraiga todas las listas de los
encabezados de columnas que se encuentran en un Panel y
úsela para extraer las listas de columnas del Panel del punto
anterior. Imprima los resultados de forma que se diferencien los
encabezados entre los dataframes en el Panel.

• Programe una función que imprima la cantidad de elementos
que tienen las bases de datos en un Panel y úsela con el Panel
del punto anterior. Imprima los resultados de forma que se
diferencien los tamaños de los diferentes dataframes en el
Panel.

• Filtre el dataframe df1 para obtener las muestras de la empresa
TAC. Determine cuantos elementos tiene dicho filtro.
3

• Determine cuantos elementos nulos hay en cada una de las
columnas del dataframe df2.


# **Revisión Teorica**

# **Estructuras de datos**
- **Tuplas**
  : Son objetos inmutables, los cuales pueden almacenar datos, pero no se pueden modificar despues de su creación. Esta estructura se usan los ( ) para definirlas.(Thakur, 2016)

- **Diccionarios**
  : En los diccionarios podemos colectar llave y valor, donde cada valor esta directamente relacionado a una llave única. En este se debe de empezar generalmente con { } en el cual se señala primero la *Key* y luego el : *Value*.(Thakur, 2016)

- **Listas**:
  Son secuencias de objetos ordenados, donde se pueden almacenar diferente tipo de datos, los cuales se deben separar con coma y se usa [ ] para definir la lista. Las listas se deben de utilizar generalmente cuando se van a almacenar varias veces el mismo objeto sean cadenas de texto o listas de números. (Thakur, 2016)

- **Series**:
  Es una estructura unidimensional de datos que se utiliza a partir de la biblioteca Pandas, adicional a esto una serie puede tener diferente tipo de datos donde se incluyen, cadenas de texto, números, booleanos, entre otros. Son parecidos a arreglos o listas. (Thakur, 2016)

-  **DataFrame**:
  Es una estructura de datos tabular que tiene columnas y filas de manera ordenada, pudiendosé ver como un grupo de series los cuales continen indices. Generalmente los DataFrames se crean a partir de la funcion de pandas.DataFrame, siendo una de las estructura óptima para el analisis de datos, manipulación de datos y adicional para visualización de datos. (Thakur, 2016)

  **Panel**
  Es una estructura de datos para datos tridimensionales, se puede utilizar sea a partir de varios DataFrames. (Thakur, 2016)

In [1]:
import pandas as pd ## Librería de pandas
from sklearn.datasets import load_breast_cancer ## Cargamos el dataset del repositorio sklearn.datasets
import numpy as np ## Librería Numpy

## Ejercicio 1

In [4]:
#Cargue en la variable df1 el conjunto de datos que se presentan en el archivo Excel, Anexo 1 - Base de Datos Excel.xlsx
df1 = pd.read_excel("Anexo 1 - Base de Datos EXCEL.xlsx", sheet_name = 1, header= 3, index_col = 0)

In [5]:
df1 ## Verficamos el DataFrame 1

Empresa Turno  Día de la semana Experiencia del conductor  \
Servicio                                                             
1            TAC     D                 5              Principiante   
2            TAD     N                 4              Principiante   
3            TAB     N                 4                   Experto   
4            TAB     D                 2                Intermedio   
5            TAC     N                 7              Principiante   
...          ...   ...               ...                       ...   
2196         TAB     D                 3                   Experto   
2197         TAB     N                 6                Intermedio   
2198         TAB     D                 3              Principiante   
2199         TAD     D                 7                Intermedio   
2200         TAD     D                 3                Intermedio   

          Formación del Paramédico Nivel del Paramédico  \
Servicio                                                  
1                                4                    I   
2                                3                    B   
3                                4                    B   
4                                3                    B   
5                                2                    I   
...                            ...                  ...   
2196                             3                    B   
2197                             4                    B   
2198                             2                    B   
2199                             1                    B   
2200                             1                    I   

          Cantidad de atenciones  Distancia recorrida (Km)  \
Servicio                                                     
1                              8                      78.4   
2                              5                      46.2   
3                              6                      62.4   
4                              7                      70.2   
5                              8                      75.6   
...                          ...                       ...   
2196                           6                      58.0   
2197                           7                      69.6   
2198                           5                      50.0   
2199                           6                      57.2   
2200                           8                      84.1   

          Velocidad máxima (Km/h)  Consumo de combustible (Galones)  \
Servicio                                                              
1                              82                               3.1   
2                              89                               1.9   
3                              99                               2.5   
4                              95                               3.0   
5                              87                               2.8   
...                           ...                               ...   
2196                          100                               1.7   
2197                           92                               3.0   
2198                           86                               2.0   
2199                           94                               2.6   
2200                           92                               3.6   

          Tiempo en Movimiento (h)  Tiempo motor ON (h)  
Servicio                                                 
1                              2.8                 3.08  
2                              2.1                 2.31  
3                              2.6                 4.68  
4                              2.6                 3.12  
5                              2.8                 3.64  
...                            ...                  ...  
2196                           2.9                 3.19  
2197                           2.9                 4.64  
2198               

In [6]:
##Cargue en la variable df2 el conjunto de datos que se presentan en el archivo CSV, Anexo 2 - Base de Datos CSV.csv
df2 = pd.read_csv("Anexo 2 - Base de Datos CSV.csv", header=0, index_col = 0, sep = ",") ## Cargamos el archivo .csv, definimos los encabezados e indicamos que debe estar separado por comas

In [7]:
df2 ## Mostramos el DataFrame 2

latitude  housing_median_age  total_rooms  total_bedrooms  \
longitude                                                              
-122.05       37.37                27.0       3885.0           661.0   
-118.30       34.26                43.0       1510.0           310.0   
-117.81       33.78                27.0       3589.0           507.0   
-118.36       33.82                28.0         67.0            15.0   
-119.67         NaN                19.0       1241.0           244.0   
...             ...                 ...          ...             ...   
-119.86         NaN                23.0       1450.0           642.0   
-118.14       34.06                27.0       5257.0          1082.0   
-119.70       36.30                10.0        956.0           201.0   
-117.12       34.10                40.0         96.0            14.0   
-119.63       34.42                42.0       1765.0           263.0   

           population  households  median_income  median_house_value  
longitude                                                             
-122.05        1537.0       606.0         6.6085            344700.0  
-118.30         809.0       277.0         3.5990            176500.0  
-117.81        1484.0       495.0         5.7934            270500.0  
-118.36          49.0        11.0         6.1359            330000.0  
-119.67         850.0       237.0         2.9375             81700.0  
...               ...         ...            ...                 ...  
-119.86        1258.0       607.0         1.1790            225000.0  
-118.14        3496.0      1036.0         3.3906            237200.0  
-119.70         693.0       220.0         2.2895             62000.0  
-117.12          46.0        14.0         3.2708            162500.0  
-119.63         753.0       260.0         8.5608            500001.0  

[3000 rows x 8 columns]

## Cargue de datos desde sklearn la base de datos de breast_cancer

In [8]:
#Cargue los datos Boston Housing Dataset que se encuentran enla variable load_boston del paquete sklearn
cancer_data =load_breast_cancer()
df3=pd.DataFrame(data=cancer_data.data, columns=cancer_data.feature_names)
df3

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0          17.99         10.38          122.80     1001.0          0.11840   
1          20.57         17.77          132.90     1326.0          0.08474   
2          19.69         21.25          130.00     1203.0          0.10960   
3          11.42         20.38           77.58      386.1          0.14250   
4          20.29         14.34          135.10     1297.0          0.10030   
..           ...           ...             ...        ...              ...   
564        21.56         22.39          142.00     1479.0          0.11100   
565        20.13         28.25          131.20     1261.0          0.09780   
566        16.60         28.08          108.30      858.1          0.08455   
567        20.60         29.33          140.10     1265.0          0.11780   
568         7.76         24.54           47.92      181.0          0.05263   

     mean compactness  mean concavity  mean concave points  mean symmetry  \
0             0.27760         0.30010              0.14710         0.2419   
1             0.07864         0.08690              0.07017         0.1812   
2             0.15990         0.19740              0.12790         0.2069   
3             0.28390         0.24140              0.10520         0.2597   
4             0.13280         0.19800              0.10430         0.1809   
..                ...             ...                  ...            ...   
564           0.11590         0.24390              0.13890         0.1726   
565           0.10340         0.14400              0.09791         0.1752   
566           0.10230         0.09251              0.05302         0.1590   
567           0.27700         0.35140              0.15200         0.2397   
568           0.04362         0.00000              0.00000         0.1587   

     mean fractal dimension  ...  worst radius  worst texture  \
0                   0.07871  ...        25.380          17.33   
1                   0.05667  ...        24.990          23.41   
2                   0.05999  ...        23.570          25.53   
3                   0.09744  ...        14.910          26.50   
4                   0.05883  ...        22.540          16.67   
..                      ...  ...           ...            ...   
564                 0.05623  ...        25.450          26.40   
565                 0.05533  ...        23.690          38.25   
566                 0.05648  ...        18.980          34.12   
567                 0.07016  ...        25.740          39.42   
568                 0.05884  ...         9.456          30.37   

     worst perimeter  worst area  worst smoothness  worst compactness  \
0             184.60      2019.0           0.16220            0.66560   
1             158.80      1956.0           0.12380            0.18660   
2             152.50      1709.0           0.14440            0.42450   
3              98.87       567.7           0.20980            0.86630   
4             152.20      1575.0           0.13740            0.20500   
..               ...         ...               ...                ...   
564           166.10      2027.0           0.14100            0.21130   
565           155.00      1731.0           0.11660            0.19220   
566           126.70      1124.0           0.11390            0.30940   
567           184.60      1821.0           0.16500            0.86810   
568            59.16       268.6           0.08996            0.06444   

     worst concavity  worst concave points  worst symmetry  \
0             0.7119                0.2654          0.4601   
1             0.2416                0.1860          0.2750   
2             0.4504                0.2430          0.3613   
3             0.6869                0.2575          0.6638   
4             0.4000                0.1625          0.2364   
..               ...                   ...             ...   
564           0.4107                0.2216         

## Creación del panel

In [9]:
##Construya un Panel con los tres (3) dataframes cargados anteriormente: df1, df2 y df3.
panel = {"DataFrame1": df1,  ## Se desarrolla un panel con los 3 DataFrames como un diccionario, con esto definirmos los keys y los values de cada DataFrame
         "DataFrame 2" : df2,
         "DataFrame 3" : df3}

In [10]:
def encabezados(panel): ## Se define la función encabezados, para luego imprimir solamente cada encabezado de las columnas de cada uno de los DataFrames
    for df_names, df in panel.items(): ## Realizamos la iteración en el Diccionario panel, donde se verifica cada item
      encabezados=df.columns ## le indicamos que utlice la función para ver todos los nombres de las columnas que tenemos.
      print("DataFrames:") ## Le ponemos nombre a cada DataFrame para mostrarlos
      print(encabezados) ## Mostramos los encabezados

In [ ]:
print(encabezados(panel))

##enca=encabezados(panel)

DataFrames:
Index(['Empresa', 'Turno', 'Día de la semana', 'Experiencia del conductor',
       'Formación del Paramédico', 'Nivel del Paramédico',
       'Cantidad de atenciones', 'Distancia recorrida (Km)',
       'Velocidad máxima (Km/h)', 'Consumo de combustible (Galones)',
       'Tiempo en Movimiento (h)', 'Tiempo motor ON (h)'],
      dtype='object')
DataFrames:
Index(['latitude', 'housing_median_age', 'total_rooms', 'total_bedrooms',
       'population', 'households', 'median_income', 'median_house_value'],
      dtype='object')
DataFrames:
Index(['mean radius', 'mean texture', 'mean perimeter', 'mean area',
       'mean smoothness', 'mean compactness', 'mean concavity',
       'mean concave points', 'mean symmetry', 'mean fractal dimension',
       'radius error', 'texture error', 'perimeter error', 'area error',
       'smoothness error', 'compactness error', 'concavity error',
       'concave points error', 'symmetry error', 'fractal dimension error',
       'worst radius', '

In [11]:
def elementos(panel): ## Definimos la función elementos
  dim={} ## creamos una variable donde se ingresarán los datos en tipo diccionario
  for df_names,df in panel.items(): ## Realizamos la iteración en el Diccionario panel, donde se verifica cada item
      dim[df_names]=df.shape ## indicamos que utilice solamente el nombre de los DataFrames y nos de sus dimensiones
  return dim

In [12]:
dim = elementos(panel) ## Variable de dimensiones

In [13]:
dim ## Mostramos las dimensiones del panel

{'DataFrame1': (2200, 12), 'DataFrame 2': (3000, 8), 'DataFrame 3': (569, 30)}

In [14]:
filtro = df1["Empresa"] == "TAC" ## Realizamos el filtro del DataFrame sobre el encabezados Empresa y usamos el indicado por el ejercicio que es TAC
filtro.sum() ## Luego esos filtros definidos de TAC los sumamos con la función .sum()

596

In [ ]:
df2.isna().sum() ## En esta línea de código llamamos la DataFrame 2 donde verificamos cuantos datos NA tiene y se suman para saber en total.

latitude              5
housing_median_age    1
total_rooms           1
total_bedrooms        1
population            3
households            2
median_income         2
median_house_value    0
dtype: int64

# **Conclusion**

- Se pudo conocer como cargar archivos desde sklearn, lo cual es muy útil para estudiar mas a fondo la ciencia de datos, adicional a esto se pudo definir funciones para el desarrollo de ciertas tareas propuestas, con esto podemos obtener una autonomia y conocimiento mas amplio de python. Se puede trabajar con varios DataFrames al tiempos creando un panel, el cual nos permite encontrar cada uno de los datos de todos los DFs utilizados, con lo cual podemos facilitar el tiempo de analisis. Al momento de definir las dimensiones pudimos observar que el DataFrame2 es el que en dimensiones de filas tiene mayor tamaño comparado a los otros con unas (3000), pero el DataFrame3 se puede ver que tiene menos filas que los otros aunque más columnas. También con la función de filtros y usando .sum() pudimos ver que de la empresa "TAC" tenemos 596 filas que contienen estas siglas en cuanto a las ambulancias, también observamos los valores que tienen NA que sde los cuales la latitude es donde hay un mayor número de esto con 5. 

# **Referencias**

- Álvarez, C. (2020). Introducción al Jupyter Notebook y aplicaciones básicas. https://repository.unad.edu.co/handle/10596/35681

- Thakur, A. (2016). Python: Real-World Data Science. Packt Publishing. https://bibliotecavirtual.unad.edu.co/login?url=https://search.ebscohost.com/login.aspx?direct=true&db=nlebk&AN=1250754&lang=es&site=ehost-live&ebv=EB&ppid=pp_199